File: app/src/main/java/com/xpf/p2p/activity/GuideActivity.java
package com.xpf.p2p.activity;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.view.ViewTreeObserver;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.viewpager.widget.ViewPager;

import com.xpf.common.cons.SpKey;
import com.xpf.common.utils.SpUtil;
import com.xpf.common.utils.UIUtils;
import com.xpf.p2p.P2PApplication;
import com.xpf.p2p.R;
import com.xpf.p2p.adapter.VpGuideAdapter;
import com.xpf.p2p.ui.login.view.LoginActivity;

import java.util.ArrayList;


/**
 * Created by xpf on 2016/11/11 :)
 * Function:引导页
 * {@link # https://github.com/xinpengfei520/P2P}
 */
public class GuideActivity extends AppCompatActivity {

    ViewPager viewPager;
    LinearLayout llPointGroup;
    ImageView ivRedPoint;
    RelativeLayout rlPoints;
    TextView tvRightAway;
    // 两点间的间距
    private int leftMarg;
    private int widthDpi;
    private ArrayList<ImageView> mImageViews;
    private final int mResIds[] = new int[]{R.drawable.guide1, R.drawable.guide2,
            R.drawable.guide3, R.drawable.guide4};

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_guide);
        viewPager = findViewById(R.id.viewPager);
        llPointGroup = findViewById(R.id.llPointGroup);
        ivRedPoint = findViewById(R.id.ivRedPoint);
        rlPoints = findViewById(R.id.rlPoints);
        tvRightAway = findViewById(R.id.tvRightAway);
        tvRightAway.setOnClickListener(v -> {
            startActivity(new Intent(GuideActivity.this, LoginActivity.class));
            finish();
        });
        initData();
        initImageViewList();
    }

    private void initImageViewList() {
        //保存图片资源的集合
        mImageViews = new ArrayList<>();
        ImageView imageView;
        //循环遍历图片资源，然后保存到集合中
        for (int resId : mResIds) {
            //添加图片到集合中
            imageView = new ImageView(this);
            imageView.setBackgroundResource(resId);
            mImageViews.add(imageView);
        }

        if (mImageViews != null && mImageViews.size() > 0) {
            viewPager.setAdapter(new VpGuideAdapter(mImageViews));
        }
    }

    /**
     * 初始化引导页的数据
     */
    private void initData() {
        widthDpi = UIUtils.dp2px(8);
        for (int i = 0; i < mResIds.length; i++) {
            // 添加灰色的点
            ImageView point = new ImageView(this);
            // 给Point设置shape为oval,即圆
            point.setImageResource(R.drawable.point_gray);
            // 设置点的大小
            LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(widthDpi, widthDpi);
            if (i != 0) {
                params.leftMargin = widthDpi; //设置间距
            }
            point.setLayoutParams(params);
            llPointGroup.addView(point);
        }

        /**
         * 求间距 构造方法--> 测量(measure--onMeasure)--> layout-->onLayout-->draw-->onDraw
         */
        ivRedPoint.getViewTreeObserver().addOnGlobalLayoutListener(new MyOnGlobalLayoutListener());
        // 监听viewPager页面滑动的百分比
        viewPager.addOnPageChangeListener(new MyOnPageChangeListener());
    }

    class MyOnPageChangeListener implements ViewPager.OnPageChangeListener {

        /**
         * @param position             当前滑动页面的下标位置
         * @param positionOffset       滑动了页面的百分比
         * @param positionOffsetPixels 滑动了页面多少像数
         */
        @Override
        public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
            // 坐标 = 起始位置 + 红点移动的距离
            float leftMargin = (position + positionOffset) * leftMarg;
            RelativeLayout.LayoutParams paramgs = (RelativeLayout.LayoutParams) ivRedPoint.getLayoutParams();
            // 距离左边的距离
            paramgs.leftMargin = (int) leftMargin;
            ivRedPoint.setLayoutParams(paramgs);
        }

        @Override
        public void onPageSelected(int position) {
            tvRightAway.setVisibility(position == 3 ? View.VISIBLE : View.GONE);
        }

        @Override
        public void onPageScrollStateChanged(int state) {

        }
    }

    class MyOnGlobalLayoutListener implements ViewTreeObserver.OnGlobalLayoutListener {

        @Override
        public void onGlobalLayout() {
            ivRedPoint.getViewTreeObserver().removeGlobalOnLayoutListener(this);
            // 间距 = 第一个点距离左边距离 - 第0个点距离左边距离
            leftMarg = llPointGroup.getChildAt(1).getLeft() - llPointGroup.getChildAt(0).getLeft();
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
        SpUtil.getInstance(P2PApplication.getContext()).save(SpKey.IS_NEED_GUIDE, false);
    }
}

File: app/src/main/java/com/xpf/p2p/activity/GuideActivity.kt
package com.xpf.p2p.activity

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.view.ViewTreeObserver
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.RelativeLayout
import androidx.appcompat.app.AppCompatActivity
import androidx.viewpager2.widget.ViewPager2
import com.xpf.p2p.App
import com.xpf.p2p.R
import com.xpf.p2p.adapter.VpGuideAdapter
import com.xpf.p2p.constants.SpKey
import com.xpf.p2p.databinding.ActivityGuideBinding
import com.xpf.p2p.ui.login.view.LoginActivity
import com.xpf.p2p.utils.SpUtil
import com.xpf.p2p.utils.StatusBarUtils
import com.xpf.p2p.utils.UIUtils

/**
 * Created by xpf on 2016/11/11 :)
 * Function:引导页
 */
class GuideActivity : AppCompatActivity() {

    private lateinit var binding: ActivityGuideBinding
    private var leftMarg: Int = 0
    private var widthDpi: Int = 0
    private val mResIds = intArrayOf(R.drawable.guide1, R.drawable.guide2, R.drawable.guide3, R.drawable.guide4)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityGuideBinding.inflate(layoutInflater)
        setContentView(binding.root)
        // 沉浸式：状态栏透明，引导图延伸到状态栏下方（图片偏彩色/深色，用浅色图标）
        StatusBarUtils.setImmersiveStatusBar(this, darkIcons = false)
        binding.tvRightAway.setOnClickListener {
            startActivity(Intent(this, LoginActivity::class.java))
            finish()
        }
        binding.viewPager.adapter = VpGuideAdapter(mResIds)
        initData()
    }

    private fun initData() {
        widthDpi = UIUtils.dp2px(8)
        for (i in mResIds.indices) {
            val point = ImageView(this)
            point.setImageResource(R.drawable.point_gray)
            val params = LinearLayout.LayoutParams(widthDpi, widthDpi)
            if (i != 0) {
                params.leftMargin = widthDpi
            }
            point.layoutParams = params
            binding.llPointGroup.addView(point)
        }

        binding.ivRedPoint.viewTreeObserver.addOnGlobalLayoutListener(object : ViewTreeObserver.OnGlobalLayoutListener {
            override fun onGlobalLayout() {
                binding.ivRedPoint.viewTreeObserver.removeOnGlobalLayoutListener(this)
                leftMarg = binding.llPointGroup.getChildAt(1).left - binding.llPointGroup.getChildAt(0).left
            }
        })

        binding.viewPager.registerOnPageChangeCallback(object : ViewPager2.OnPageChangeCallback() {
            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
                val leftMargin = (position + positionOffset) * leftMarg
                val params = binding.ivRedPoint.layoutParams as RelativeLayout.LayoutParams
                params.leftMargin = leftMargin.toInt()
                binding.ivRedPoint.layoutParams = params
            }

            override fun onPageSelected(position: Int) {
                binding.tvRightAway.visibility = if (position == mResIds.size - 1) View.VISIBLE else View.GONE
            }
        })
    }

    override fun onResume() {
        super.onResume()
        SpUtil.getInstance(App.context).save(SpKey.IS_NEED_GUIDE, false)
    }
}
